File: Platforms/Windows/Reign.Video.D3D9/RasterizerState.cpp
#include "pch.h"
#include "RasterizerState.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region Constructor
	RasterizerState::RasterizerState(DisposableI^ parent, RasterizerStateDescI^ desc)
	: Disposable(parent)
	{
		video = parent->FindParentOrSelfWithException<Video^>();
		this->desc = (RasterizerStateDesc^)desc;
	}

	RasterizerState::~RasterizerState()
	{
		disposeChilderen();
	}
	#pragma endregion
	
	#pragma region Methods
	void RasterizerState::Enable()
	{
		video->Device->SetRenderState(D3DRS_FILLMODE, desc->FillMode);
		video->Device->SetRenderState(D3DRS_CULLMODE, desc->CullMode);
		video->Device->SetRenderState(D3DRS_MULTISAMPLEANTIALIAS, desc->MultisampleEnable);
	}
	#pragma endregion
}
}
}

File: Platforms/Common/Reign.Video.XNA/States/RasterizerState.cs
using Reign.Core;
using X = Microsoft.Xna.Framework.Graphics;

namespace Reign.Video.XNA
{
	public class RasterizerState : Disposable, RasterizerStateI
	{
		#region Properties
		private Video video;
		private RasterizerStateDesc desc;
		#endregion

		#region Constructors
		public static RasterizerState New(DisposableI parent, RasterizerStateDescI desc)
		{
			return new RasterizerState(parent, desc);
		}

		public RasterizerState(DisposableI parent, RasterizerStateDescI desc)
		: base(parent)
		{
			this.video = parent.FindParentOrSelfWithException<Video>();
			this.desc = (RasterizerStateDesc)desc;
		}
		#endregion

		#region Methods
		public void Enable()
		{
			var rasterizerState = new X.RasterizerState()
			{
				CullMode = desc.cullMode,
				FillMode = desc.fillMode,
				MultiSampleAntiAlias = desc.multisampleEnable
			};

			video.Device.RasterizerState = rasterizerState;
		}
		#endregion
	}
}